### ISSUE TITLE
Add support for relative paths

### ISSUE THREAD

**Issue by nik875** [toebeann/gib#11]
![Screenshot 2024-08-18 at 12 35 39 PM](https://github.com/user-attachments/assets/3225bc6c-fa65-4baf-9aa2-409661de7a0e)

Not sure what's going on here, it's a very vague error. This happened as I was running gib once it reached the test phase. I am running on M1 Mac with the latest Subnautica version. I had also attempted to create the experimental vanilla game shortcut, and this gets a bit further by opening a window but eventually causes the same cryptic error.

I am also a developer so I can lend a hand in fixing this if someone points me in the right direction, but as of now the error is very undescriptive and I'm not sure where to start.

**Comment by nik875** [comment]
It would also be nice for official uninstallation instructions to be posted detailing how to revert to the Vanilla state before gib touched anything.

**Comment by toebeann** [comment]
Unknown error being vague is a steam thing, usually means the launch options aren't correct. If I were you, try gib again from square one as today I released an update that streamlines everything, sets up the launch options for you, and adds the option to create a shortcut to launch the game vanilla (though with Subnautica I believe that vanilla shortcut doesn't work, limitation with the game itself unfortunately)

As for instructions on uninstall, yes this is planned. Unfortunately my free time isn't infinite xD

**Comment by nik875** [comment]
I didn't touch the launch options at all, just entered the paths and proceeded through the install. This was on the latest version of gib (to the best of my knowledge, I used the curl command in the README). I see the last commit was about an hour ago and I ran it after that time, so I assume the version was updated. I will go ahead and try again just in case.

I've also tried reinstalling the game and fully deleting the Subnautica directory in Steam/steamapps/common, but neither fixes the game and reverts to Vanilla state. Could you let me know how I could get it working again in Vanilla as well?

**Comment by toebeann** [comment]
I edited my previous comment to give more info. Unsure how/why your launch options ended up incorrect, been testing with Subnautica myself. You on M-series chip? Can you copy and paste the steam launch options?

**Comment by toebeann** [comment]
Shouldn't matter where you ran the command, I've ran it in a variety of locations myself. It is supposed to copy the files over, not symlink. I'm on intel, might be a quirk I've not encountered. Can't really test without an M-series Mac unfortunately, but I'll spend some time looking through docs etc and see what I can do

**Comment by nik875** [comment]
You're right, it was copied and not symlinked, sorry about that.

It seems to be the full path actually, not just the Downloads folder:

```
./Subnautica
├── Subnautica.app
│   ├── Contents
│   │   ├── Frameworks
│   │   ├── Info.plist
│   │   ├── MacOS
│   │   ├── MonoBleedingEdge
│   │   ├── Plugins
│   │   └── Resources
│   └── SNAppData
│       ├── SavedGames
│       └── SavedGamesBackup
└── Users
    └── NikhilKalidasu
        └── Downloads

14 directories, 1 file
```

**Comment by toebeann** [comment]
So it just created a Users/X/Downloads folder in Subnautica rather than copying anything over? Very strange...

**Comment by nik875** [comment]
Path to run_bepinex.sh: ./run_bepinex.sh

Path to Unity game app: /Users/NikhilKalidasu/Library/Application Support/Steam/steamapps/common/Subnautica/Subnautica.app

**Comment by nik875** [comment]
![image](https://github.com/user-attachments/assets/d0f8f557-b189-462b-84f6-96a2789738b6)

I changed the path in the launch options to go down the full directory path to the copied over run_bepinex.sh and got this error also.

**Comment by toebeann** [comment]
> Path to run_bepinex.sh: ./run_bepinex.sh

Don't use a relative path please, hasn't been tested. Just highlight the file and cmd+c it, then cmd+v in terminal. If it doesn't accept that, opt+cmd+c.

**Comment by toebeann** [comment]
> ![image](https://github.com/user-attachments/assets/d0f8f557-b189-462b-84f6-96a2789738b6)
> 
> I changed the path in the launch options to go down the full directory path to the copied over run_bepinex.sh and got this error also.

Yeah I didn't suggest trying to workaround it for a reason xD, the things gib has to take care of under the hood to make apple happy is a lot more than just copying the files over

**Comment by nik875** [comment]
Yep using the absolute path fixed the problem, thanks! It would be nice to have that clarification in the README.

The only remaining issue now is that when I try running the game with BepInEx, it has been stuck on a black screen for the past several minutes. About how long does it normally take to start up?

Although this is more of an issue with the mod and not gib, so understandable if you want to close this issue now.

**Comment by toebeann** [comment]
Let me know if using absolute paths works for you and if so then I'll chalk this up to a quirk of my input validation not rejecting relative paths but my processing path not handling relative paths correctly, and try to fix this in the next few days

If using absolute paths doesn't resolve the issue, let me know

**Comment by toebeann** [comment]
> The only remaining issue now is that when I try running the game with BepInEx, it has been stuck on a black screen for the past several minutes. About how long does it normally take to start up?
> 
> Although this is more of an issue with the mod and not gib, so understandable if you want to close this issue now.

Yeah when you start up with just BepInEx it's very fast for me. So that will be whatever mods you have, and that will vary from mod to mod. Could also be that the mod itself just doesn't support macOS, not all will "just work" on macOS, depends on the mod.

**Comment by nik875** [comment]
The only mod should be BepInEx itself, so the issue is definitely there.

**Comment by toebeann** [comment]
> Yep using the absolute path fixed the problem, thanks! It would be nice to have that clarification in the README.

Thanks for letting me know, yes I'll add a caveat about relative paths until I have time to fix it, but be aware it's a v0 (i.e. still a work in progress) and deviating from the instructions is bound to cause issues. If you deviated from the instructions and something didn't work, your first thought as a user - especially a developer! - should be to try again, doing it exactly the way the instructions describe ;)

**Comment by toebeann** [comment]
Oh, and I'll leave this issue open until I've fixed the relative paths issue :)

**Comment by toebeann** [comment]
> The only mod should be BepInEx itself, so the issue is definitely there.

Might be an M-series issue as it's having to emulate intel then, running everything through Rosetta. For me on intel - which is what BepInEx was made for - it's very fast indeed.

**Comment by nik875** [comment]
Agreed that it's probably an M-series issue. However the Vanilla shortcut has the same problem. It gets a bit further and I see the opening photosensitivity warning before it returns to a black screen and gets stuck. Ideally the Vanilla shortcut should not be using any mods at all including BepInEx, so not sure what's going on there.

**Comment by toebeann** [comment]
> Agreed that it's probably an M-series issue. However the Vanilla shortcut has the same problem. It gets a bit further and I see the opening photosensitivity warning before it returns to a black screen and gets stuck. Ideally the Vanilla shortcut should not be using any mods at all including BepInEx, so not sure what's going on there.

I mentioned earlier in this thread that the vanilla shortcut doesn't work for Subnautica - it's a game-specific issue. The vanilla shortcut is experimental and may not work for all Steam games, as gib will have told you :)

**Comment by toebeann** [comment]
Some more info:

The way the vanilla shortcut works is to simply run the game's .app directly, without running the run_bepinex.sh script. Doing this means we're not launching the game through Steam, so the Steamworks API doesn't get injected into the game, and some games - including Subnautica - close and restart the game when they detect this. Unfortunately, this leads to it launching the game directly with Steam (with launch options), which means it runs the run_bepinex.sh script and injects BepInEx.

**Comment by toebeann** [comment]
@nik875 Re: vanilla shortcut for Steam games, I just wanted to let you know that I have just released gib v0.3.4 which now uses an improved methodology for providing the vanilla shortcut, making use of the `--doorstop_enabled` flag used in BepInEx packs which use doorstop v4 or newer. I have also updated my BepInEx packs for Subnautica and Subnautica: Below Zero to use doorstop v4.

So, if you want to make use of the vanilla shortcut functionality I advise you to:
1. Remove any vanilla shortcuts from Steam that were created previously (I have not yet added functionality for gib to update the shortcuts if they already exist)
2. [Download the latest version of my BepInEx pack for Subnautica](https://www.nexusmods.com/subnautica/mods/1108)
3. Run gib: `curl -fsSL https://cdn.jsdelivr.net/gh/toebeann/gib/gib.sh | bash`
4. Press `y` then `Enter` when asked if you would like a vanilla shortcut 😄

Do note that I still haven't had time to look into the relative path issue, so make sure to use absolute paths!
